File: app/src/main/java/com/adasa/foodplan/data/repository/SettingsRepository.kt
package com.adasa.foodplan.data.repository

import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.*
import com.adasa.foodplan.data.local.dao.OptimizerSettingsDao
import com.adasa.foodplan.data.local.entity.BatchCookingGroupEntity
import com.adasa.foodplan.data.local.entity.MealSlotConfigEntity
import com.adasa.foodplan.data.local.entity.toEntity
import com.adasa.foodplan.domain.model.*
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map
import kotlinx.datetime.DayOfWeek
import kotlinx.datetime.isoDayNumber
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class SettingsRepository @Inject constructor(
    private val dataStore: DataStore<androidx.datastore.preferences.core.Preferences>,
    private val dao: OptimizerSettingsDao
) {
    companion object {
        // Nutrition goals
        val KCAL_TARGET = doublePreferencesKey("kcal_target")
        val PROTEIN_TARGET = doublePreferencesKey("protein_target")
        val FAT_TARGET = doublePreferencesKey("fat_target")
        val CARBS_TARGET = doublePreferencesKey("carbs_target")
        val AUTO_FIELD = stringPreferencesKey("auto_macro_field")
        val MIN_KCAL = doublePreferencesKey("min_kcal_per_day")
        val MAX_KCAL = doublePreferencesKey("max_kcal_per_day")
        val MIN_PROTEIN = doublePreferencesKey("min_protein_per_day")
        val MAX_PROTEIN = doublePreferencesKey("max_protein_per_day")
        val MIN_FAT = doublePreferencesKey("min_fat_per_day")
        val MAX_FAT = doublePreferencesKey("max_fat_per_day")
        val MIN_CARBS = doublePreferencesKey("min_carbs_per_day")
        val MAX_CARBS = doublePreferencesKey("max_carbs_per_day")
        // Diet
        val DIET_TYPES = stringPreferencesKey("diet_types")
        val ALLERGIES = stringPreferencesKey("allergies")
        val EXCLUDED_INGREDIENT_IDS = stringPreferencesKey("excluded_ingredient_ids")
        val PREFERRED_INGREDIENT_IDS = stringPreferencesKey("preferred_ingredient_ids")
        val DISLIKED_INGREDIENT_IDS = stringPreferencesKey("disliked_ingredient_ids")
        // Variety
        val MAX_DAYS_IN_A_ROW = intPreferencesKey("max_days_in_a_row")
        val UNIQUE_WEEKS = intPreferencesKey("unique_weeks")
        val PROTEIN_VARIETY = booleanPreferencesKey("protein_source_variety")
        // Schedule
        val SNACK_OPTIONAL_FILL = booleanPreferencesKey("snack_optional_fill")
        // Shopping
        val SHOPPING_DAYS = stringPreferencesKey("shopping_days")
        val SHOPPING_INTERVAL = intPreferencesKey("shopping_interval_weeks")
        // Protein powder
        val POWDER_INGREDIENT_ID = stringPreferencesKey("powder_ingredient_id")
        val POWDER_NAME = stringPreferencesKey("powder_name")
        val POWDER_PROTEIN_PER_100G = doublePreferencesKey("powder_protein_per_100g")
        val POWDER_KCAL_PER_100G = doublePreferencesKey("powder_kcal_per_100g")
        val POWDER_GRAMS_IN_STOCK = doublePreferencesKey("powder_grams_in_stock")
        val POWDER_AUTO_FILL = booleanPreferencesKey("powder_auto_fill")
        val POWDER_LOW_STOCK_WARNING = booleanPreferencesKey("powder_low_stock_warning")
        // App
        val THEME = stringPreferencesKey("app_theme")
    }

    // ── Full config assembly ───────────────────────────────────────────────

    suspend fun getMealPlanConfig(): MealPlanConfig {
        val prefs = dataStore.data.first()
        val slotConfigs = dao.getAllMealSlotConfigsOnce()
        val batchGroups = dao.getAllBatchGroupsOnce()
        val rules = dao.getAllRulesOnce()

        return MealPlanConfig(
            schedule = buildScheduleConfig(prefs, slotConfigs, batchGroups),
            goals = buildNutritionGoals(prefs),
            diet = buildDietPreferences(prefs),
            rules = rules.map { it.toDomain() },
            variety = buildVarietyConfig(prefs),
            proteinPowder = buildProteinPowder(prefs),
            shopping = buildShoppingConfig(prefs)
        )
    }

    fun getMealPlanConfigFlow(): Flow<MealPlanConfig> =
        combine(
            dataStore.data,
            dao.getAllMealSlotConfigs(),
            dao.getAllBatchGroups(),
            dao.getAllRules()
        ) { prefs, slots, batches, rules ->
            MealPlanConfig(
                schedule = buildScheduleConfig(prefs, slots, batches),
                goals = buildNutritionGoals(prefs),
                diet = buildDietPreferences(prefs),
                rules = rules.map { it.toDomain() },
                variety = buildVarietyConfig(prefs),
                proteinPowder = buildProteinPowder(prefs),
                shopping = buildShoppingConfig(prefs)
            )
        }

    // ── Builders ──────────────────────────────────────────────────────────

    private fun buildScheduleConfig(
        prefs: androidx.datastore.preferences.core.Preferences,
        slots: List<MealSlotConfigEntity>,
        batches: List<BatchCookingGroupEntity>
    ): MealScheduleConfig {
        val slotMap = slots.associate { entity ->
            DayOfWeek(entity.dayOfWeek) to entity.toDomain()
        }.let { map ->
            DayOfWeek.entries.associateWith { map[it] ?: defaultDayConfig(it) }
        }
        return MealScheduleConfig(
            mealSlots = slotMap,
            batchGroups = batches.map { it.toDomain() },
            snackOptionalFill = prefs[SNACK_OPTIONAL_FILL] ?: true
        )
    }

    private fun buildNutritionGoals(prefs: androidx.datastore.preferences.core.Preferences) =
        NutritionGoals(
            kcalTarget = prefs[KCAL_TARGET] ?: 1450.0,
            proteinTarget = prefs[PROTEIN_TARGET],
            fatTarget = prefs[FAT_TARGET],
            carbsTarget = prefs[CARBS_TARGET],
            autoField = prefs[AUTO_FIELD]?.let { MacroField.valueOf(it) } ?: MacroField.PROTEIN,
            minKcalPerDay = prefs[MIN_KCAL],
            maxKcalPerDay = prefs[MAX_KCAL],
            minProteinPerDay = prefs[MIN_PROTEIN],
            maxProteinPerDay = prefs[MAX_PROTEIN],
            minFatPerDay = prefs[MIN_FAT],
            maxFatPerDay = prefs[MAX_FAT],
            minCarbsPerDay = prefs[MIN_CARBS],
            maxCarbsPerDay = prefs[MAX_CARBS]
        )

    private fun buildDietPreferences(prefs: androidx.datastore.preferences.core.Preferences) =
        DietPreferences(
            dietTypes = prefs[DIET_TYPES]?.splitToSet()
                ?.mapNotNull { runCatching { DietType.valueOf(it) }.getOrNull() }
                ?.toSet() ?: emptySet(),
            allergies = prefs[ALLERGIES]?.splitToSet()
                ?.mapNotNull { runCatching { AllergyType.valueOf(it) }.getOrNull() }
                ?.toSet() ?: emptySet(),
            excludedIngredientIds = prefs[EXCLUDED_INGREDIENT_IDS]?.splitToSet() ?: emptySet(),
            preferredIngredientIds = prefs[PREFERRED_INGREDIENT_IDS]?.splitToSet() ?: emptySet(),
            dislikedIngredientIds = prefs[DISLIKED_INGREDIENT_IDS]?.splitToSet() ?: emptySet()
        )

    private fun buildVarietyConfig(prefs: androidx.datastore.preferences.core.Preferences) =
        VarietyConfig(
            maxDaysInARow = prefs[MAX_DAYS_IN_A_ROW] ?: 2,
            uniqueWeeksBeforeRepeat = prefs[UNIQUE_WEEKS] ?: 3,
            proteinSourceVariety = prefs[PROTEIN_VARIETY] ?: true
        )

    private fun buildProteinPowder(prefs: androidx.datastore.preferences.core.Preferences): ProteinPowder? {
        val id   = prefs[POWDER_INGREDIENT_ID] ?: return null
        val name = prefs[POWDER_NAME] ?: return null
        return ProteinPowder(
            ingredientId   = id,
            name           = name,
            proteinPer100g = prefs[POWDER_PROTEIN_PER_100G] ?: 0.0,
            kcalPer100g    = prefs[POWDER_KCAL_PER_100G]    ?: 0.0,
            gramsInStock   = prefs[POWDER_GRAMS_IN_STOCK]   ?: 0.0,
            autoFillGap    = prefs[POWDER_AUTO_FILL]         ?: true,
            lowStockWarning = prefs[POWDER_LOW_STOCK_WARNING] ?: true
        )
    }

    private fun buildShoppingConfig(prefs: androidx.datastore.preferences.core.Preferences) =
        ShoppingConfig(
            shoppingDays = prefs[SHOPPING_DAYS]?.splitToSet()
                ?.mapNotNull { runCatching { DayOfWeek(it.toInt()) }.getOrNull() }
                ?.toSet() ?: setOf(DayOfWeek.SUNDAY),
            intervalWeeks = prefs[SHOPPING_INTERVAL] ?: 1
        )

    private fun defaultDayConfig(day: DayOfWeek) = DayMealConfig(
        breakfast = false,
        lunch = true,
        dinner = true,
        snackCount = if (day == DayOfWeek.FRIDAY || day == DayOfWeek.SATURDAY) 1 else 0
    )

    // ── Individual setters ────────────────────────────────────────────────

    suspend fun setKcalTarget(kcal: Double) = dataStore.edit { it[KCAL_TARGET] = kcal }
    suspend fun setProteinTarget(g: Double?) = dataStore.edit {
        if (g != null) it[PROTEIN_TARGET] = g else it.remove(PROTEIN_TARGET)
    }
    suspend fun setFatTarget(g: Double?) = dataStore.edit {
        if (g != null) it[FAT_TARGET] = g else it.remove(FAT_TARGET)
    }
    suspend fun setCarbsTarget(g: Double?) = dataStore.edit {
        if (g != null) it[CARBS_TARGET] = g else it.remove(CARBS_TARGET)
    }
    suspend fun setAutoField(field: MacroField) = dataStore.edit { it[AUTO_FIELD] = field.name }
    suspend fun setMinKcal(v: Double?) = dataStore.edit { if (v != null) it[MIN_KCAL] = v else it.remove(MIN_KCAL) }
    suspend fun setMaxKcal(v: Double?) = dataStore.edit { if (v != null) it[MAX_KCAL] = v else it.remove(MAX_KCAL) }
    suspend fun setMinProtein(v: Double?) = dataStore.edit { if (v != null) it[MIN_PROTEIN] = v else it.remove(MIN_PROTEIN) }
    suspend fun setMaxProtein(v: Double?) = dataStore.edit { if (v != null) it[MAX_PROTEIN] = v else it.remove(MAX_PROTEIN) }
    suspend fun setMinFat(v: Double?) = dataStore.edit { if (v != null) it[MIN_FAT] = v else it.remove(MIN_FAT) }
    suspend fun setMaxFat(v: Double?) = dataStore.edit { if (v != null) it[MAX_FAT] = v else it.remove(MAX_FAT) }
    suspend fun setMinCarbs(v: Double?) = dataStore.edit { if (v != null) it[MIN_CARBS] = v else it.remove(MIN_CARBS) }
    suspend fun setMaxCarbs(v: Double?) = dataStore.edit { if (v != null) it[MAX_CARBS] = v else it.remove(MAX_CARBS) }

    suspend fun setDietTypes(types: Set<DietType>) = dataStore.edit {
        it[DIET_TYPES] = types.joinToString(",") { t -> t.name }
    }
    suspend fun setAllergies(allergies: Set<AllergyType>) = dataStore.edit {
        it[ALLERGIES] = allergies.joinToString(",") { a -> a.name }
    }
    suspend fun setExcludedIngredients(ids: Set<String>) = dataStore.edit {
        it[EXCLUDED_INGREDIENT_IDS] = ids.joinToString(",")
    }
    suspend fun setPreferredIngredients(ids: Set<String>) = dataStore.edit {
        it[PREFERRED_INGREDIENT_IDS] = ids.joinToString(",")
    }
    suspend fun setDislikedIngredients(ids: Set<String>) = dataStore.edit {
        it[DISLIKED_INGREDIENT_IDS] = ids.joinToString(",")
    }

    suspend fun setVariety(config: VarietyConfig) = dataStore.edit {
        it[MAX_DAYS_IN_A_ROW] = config.maxDaysInARow
        it[UNIQUE_WEEKS] = config.uniqueWeeksBeforeRepeat
        it[PROTEIN_VARIETY] = config.proteinSourceVariety
    }

    suspend fun setSnackOptionalFill(enabled: Boolean) = dataStore.edit {
        it[SNACK_OPTIONAL_FILL] = enabled
    }

    suspend fun setShoppingDays(days: Set<DayOfWeek>) = dataStore.edit {
        it[SHOPPING_DAYS] = days.joinToString(",") { d -> d.isoDayNumber.toString() }
    }
    suspend fun setShoppingInterval(weeks: Int) = dataStore.edit {
        it[SHOPPING_INTERVAL] = weeks
    }

    suspend fun setProteinPowder(
        ingredientId:   String,
        name:           String,
        proteinPer100g: Double,
        kcalPer100g:    Double,
        gramsInStock:   Double
    ) = dataStore.edit {
        it[POWDER_INGREDIENT_ID]   = ingredientId
        it[POWDER_NAME]            = name
        it[POWDER_PROTEIN_PER_100G] = proteinPer100g
        it[POWDER_KCAL_PER_100G]   = kcalPer100g
        it[POWDER_GRAMS_IN_STOCK]  = gramsInStock
    }

    suspend fun restockPowder(grams: Double) = dataStore.edit {
        it[POWDER_GRAMS_IN_STOCK] = grams
    }
    suspend fun setPowderAutoFill(enabled: Boolean) = dataStore.edit { it[POWDER_AUTO_FILL] = enabled }
    suspend fun setPowderLowStockWarning(enabled: Boolean) = dataStore.edit { it[POWDER_LOW_STOCK_WARNING] = enabled }

    suspend fun setTheme(theme: String) = dataStore.edit { it[THEME] = theme }
    fun getThemeFlow(): Flow<String> = dataStore.data.map { it[THEME] ?: "system" }

    suspend fun setMealSlotConfig(dayOfWeek: DayOfWeek, config: com.adasa.foodplan.domain.model.DayMealConfig) =
        dao.upsertMealSlotConfig(
            MealSlotConfigEntity(
                dayOfWeek  = dayOfWeek.isoDayNumber,
                breakfast  = config.breakfast,
                lunch      = config.lunch,
                dinner     = config.dinner,
                snackCount = config.snackCount
            )
        )

    suspend fun saveBatchGroups(groups: List<BatchCookingGroup>) {
        dao.deleteAllBatchGroups()
        groups.forEachIndexed { i, g ->
            dao.upsertBatchGroup(g.toEntity("batch_$i"))
        }
    }

    suspend fun upsertRule(rule: OptimizerRule) =
        dao.upsertRule(rule.copy(id = rule.id.ifEmpty { UUID.randomUUID().toString() }).toEntity())

    suspend fun deleteRule(ruleId: String) = dao.deleteRuleById(ruleId)

    fun getRulesFlow(): Flow<List<OptimizerRule>> =
        dao.getAllRules().map { it.map { e -> e.toDomain() } }

    // ── Helpers ───────────────────────────────────────────────────────────

    private fun String.splitToSet(): Set<String> =
        split(",").map { it.trim() }.filter { it.isNotEmpty() }.toSet()
}

File: flutter-app/lib/data/repository/settings_repository.dart
import 'dart:async';

import 'package:drift/drift.dart';
import 'package:foodplan/data/local/app_database.dart';
import 'package:foodplan/domain/model/meal_plan_config.dart';
import 'package:foodplan/domain/model/recipe.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'package:uuid/uuid.dart';

// ── SharedPreferences keys ─────────────────────────────────────────────────

const _kKcalTarget = 'kcal_target';
const _kProteinTarget = 'protein_target';
const _kFatTarget = 'fat_target';
const _kCarbsTarget = 'carbs_target';
const _kAutoField = 'auto_macro_field';
const _kMinKcal = 'min_kcal_per_day';
const _kMaxKcal = 'max_kcal_per_day';
const _kMinProtein = 'min_protein_per_day';
const _kMaxProtein = 'max_protein_per_day';
const _kMinFat = 'min_fat_per_day';
const _kMaxFat = 'max_fat_per_day';
const _kMinCarbs = 'min_carbs_per_day';
const _kMaxCarbs = 'max_carbs_per_day';
const _kDietTypes = 'diet_types';
const _kAllergies = 'allergies';
const _kExcludedIds = 'excluded_ingredient_ids';
const _kPreferredIds = 'preferred_ingredient_ids';
const _kDislikedIds = 'disliked_ingredient_ids';
const _kVarietyLevel = 'variety_level';
const _kVarietyLunchDinnerShared = 'variety_lunch_dinner_shared';
const _kVarietyBreakfastSnackShared = 'variety_breakfast_snack_shared';
const _kVarietyLunchDinnerMustDiffer = 'variety_lunch_dinner_must_differ';
const _kVarietyProteinSource = 'protein_source_variety';
const _kVarietyBreakfastMaxWeek = 'variety_breakfast_max_week';
const _kVarietyBreakfastMaxConsec = 'variety_breakfast_max_consec';
const _kVarietyLunchMaxWeek = 'variety_lunch_max_week';
const _kVarietyLunchMaxConsec = 'variety_lunch_max_consec';
const _kVarietyDinnerMaxWeek = 'variety_dinner_max_week';
const _kVarietyDinnerMaxConsec = 'variety_dinner_max_consec';
const _kVarietySnackMaxWeek = 'variety_snack_max_week';
const _kVarietySnackMaxConsec = 'variety_snack_max_consec';
const _kSnackOptionalFill = 'snack_optional_fill';
const _kShoppingDays = 'shopping_days';
const _kShoppingInterval = 'shopping_interval_weeks';
const _kPowderIngredientId = 'powder_ingredient_id';
const _kPowderName = 'powder_name';
const _kPowderProteinPer100g = 'powder_protein_per_100g';
const _kPowderKcalPer100g = 'powder_kcal_per_100g';
const _kPowderGramsInStock = 'powder_grams_in_stock';
const _kPowderAutoFill = 'powder_auto_fill';
const _kPowderLowStockWarning = 'powder_low_stock_warning';
const _kTheme = 'app_theme';
const _kUnlimited = -1;

class SettingsRepository {
  SettingsRepository({
    required SharedPreferences prefs,
    required SettingsDao dao,
  })  : _prefs = prefs,
        _dao = dao;

  final SharedPreferences _prefs;
  final SettingsDao _dao;
  final _uuid = const Uuid();

  // ── Config assembly ───────────────────────────────────────────────────────

  Future<MealPlanConfig> getMealPlanConfig() async {
    final slotRows = await _dao.getAllMealSlotConfigsOnce();
    final batchRows = await _dao.getAllBatchCookingGroupsOnce();
    final ruleRows = await _dao.getAllOptimizerRulesOnce();
    return MealPlanConfig(
      schedule: _buildScheduleConfig(slotRows, batchRows),
      goals: _buildNutritionGoals(),
      diet: _buildDietPreferences(),
      variety: _buildVarietyConfig(),
      rules: ruleRows.map(_optimizerRuleFromRow).toList(),
      proteinPowder: _buildProteinPowder(),
      shopping: _buildShoppingConfig(),
    );
  }

  /// Stream that emits a new [MealPlanConfig] whenever prefs or DB tables change.
  Stream<MealPlanConfig> watchMealPlanConfig() {
    // Combine the three DB streams; prefs changes are handled by the broadcast
    // controller that is triggered on every setter call.
    final controller = StreamController<MealPlanConfig>.broadcast();

    // Emit on DB changes
    late StreamSubscription slotSub;
    late StreamSubscription batchSub;
    late StreamSubscription ruleSub;

    Future<void> emit() async {
      try {
        final config = await getMealPlanConfig();
        if (!controller.isClosed) controller.add(config);
      } catch (e) {
        if (!controller.isClosed) controller.addError(e);
      }
    }

    slotSub = _dao.watchMealSlotConfigs().listen((_) => emit());
    batchSub = _dao.watchBatchCookingGroups().listen((_) => emit());
    ruleSub = _dao.watchOptimizerRules().listen((_) => emit());

    // Emit immediately
    emit();

    controller.onCancel = () {
      slotSub.cancel();
      batchSub.cancel();
      ruleSub.cancel();
    };

    return controller.stream;
  }

  // ── Builders ──────────────────────────────────────────────────────────────

  MealScheduleConfig _buildScheduleConfig(
    List<MealSlotConfigsTableData> slots,
    List<BatchCookingGroupsTableData> batches,
  ) {
    final slotMap = {
      for (final row in slots)
        row.dayOfWeek: DaySlotConfig(
          hasBreakfast: row.hasBreakfast,
          hasLunch: row.hasLunch,
          hasDinner: row.hasDinner,
          snackCount: row.snackCount,
        ),
    };
    // Fill in defaults for missing days
    final perDay = {
      for (int d = 1; d <= 7; d++)
        d: slotMap[d] ??
            DaySlotConfig(
              hasBreakfast: false,
              hasLunch: true,
              hasDinner: true,
              snackCount: (d == 5 || d == 6) ? 1 : 0,
            ),
    };
    return MealScheduleConfig(
      perDay: perDay,
      batchCookingGroups: batches
          .map((row) => BatchCookingGroup(
                category: MealCategory.fromFirestore(row.category) ??
                    MealCategory.lunch,
                days: row.days.isEmpty
                    ? {}
                    : row.days
                        .split(',')
                        .map((s) => int.tryParse(s.trim()) ?? 1)
                        .toSet(),
                batchNumber: row.batchNumber,
              ))
          .toList(),
      snackOptionalFill: _prefs.getBool(_kSnackOptionalFill) ?? true,
    );
  }

  NutritionGoals _buildNutritionGoals() => NutritionGoals(
        kcalTarget: _prefs.getDouble(_kKcalTarget) ?? 1450.0,
        proteinTarget: _prefs.getDouble(_kProteinTarget),
        fatTarget: _prefs.getDouble(_kFatTarget),
        carbsTarget: _prefs.getDouble(_kCarbsTarget),
        autoField:
            MacroField.fromFirestore(_prefs.getString(_kAutoField)),
        minKcalPerDay: _prefs.getDouble(_kMinKcal),
        maxKcalPerDay: _prefs.getDouble(_kMaxKcal),
        minProteinPerDay: _prefs.getDouble(_kMinProtein),
        maxProteinPerDay: _prefs.getDouble(_kMaxProtein),
        minFatPerDay: _prefs.getDouble(_kMinFat),
        maxFatPerDay: _prefs.getDouble(_kMaxFat),
        minCarbsPerDay: _prefs.getDouble(_kMinCarbs),
        maxCarbsPerDay: _prefs.getDouble(_kMaxCarbs),
      );

  DietPreferences _buildDietPreferences() => DietPreferences(
        dietTypes: _splitToSet(_prefs.getString(_kDietTypes))
            .map((s) {
              try {
                return DietType.values.firstWhere(
                  (e) => e.firestoreName == s.toUpperCase(),
                );
              } catch (_) {
                return null;
              }
            })
            .whereType<DietType>()
            .toSet(),
        allergies: _splitToSet(_prefs.getString(_kAllergies))
            .map((s) {
              try {
                return AllergyType.values.firstWhere(
                  (e) => e.firestoreName == s.toUpperCase(),
                );
              } catch (_) {
                return null;
              }
            })
            .whereType<AllergyType>()
            .toSet(),
        excludedIngredientIds: _splitToSet(_prefs.getString(_kExcludedIds)),
        preferredIngredientIds:
            _splitToSet(_prefs.getString(_kPreferredIds)),
        dislikedIngredientIds: _splitToSet(_prefs.getString(_kDislikedIds)),
      );

  VarietyConfig _buildVarietyConfig() {
    int? toLimit(String key) {
      final v = _prefs.getInt(key);
      return (v == null || v == _kUnlimited) ? null : v;
    }

    return VarietyConfig(
      level: () {
        try {
          return VarietyLevel.values.firstWhere(
            (e) =>
                e.name.toUpperCase() ==
                (_prefs.getString(_kVarietyLevel) ?? '').toUpperCase(),
          );
        } catch (_) {
          return VarietyLevel.balanced;
        }
      }(),
      lunchDinnerSharedRecency:
          _prefs.getBool(_kVarietyLunchDinnerShared) ?? true,
      breakfastSnackSharedRecency:
          _prefs.getBool(_kVarietyBreakfastSnackShared) ?? false,
      lunchDinnerMustDiffer:
          _prefs.getBool(_kVarietyLunchDinnerMustDiffer) ?? true,
      proteinSourceVariety:
          _prefs.getBool(_kVarietyProteinSource) ?? true,
      perCategory: {
        MealCategory.breakfast: MealCategoryVariety(
          maxTimesPerWeek: toLimit(_kVarietyBreakfastMaxWeek),
          maxConsecutiveDays: toLimit(_kVarietyBreakfastMaxConsec),
        ),
        MealCategory.lunch: MealCategoryVariety(
          maxTimesPerWeek:
              _prefs.containsKey(_kVarietyLunchMaxWeek)
                  ? toLimit(_kVarietyLunchMaxWeek)
                  : 3,
          maxConsecutiveDays:
              _prefs.containsKey(_kVarietyLunchMaxConsec)
                  ? toLimit(_kVarietyLunchMaxConsec)
                  : 2,
        ),
        MealCategory.dinner: MealCategoryVariety(
          maxTimesPerWeek:
              _prefs.containsKey(_kVarietyDinnerMaxWeek)
                  ? toLimit(_kVarietyDinnerMaxWeek)
                  : 2,
          maxConsecutiveDays:
              _prefs.containsKey(_kVarietyDinnerMaxConsec)
                  ? toLimit(_kVarietyDinnerMaxConsec)
                  : 2,
        ),
        MealCategory.snack: MealCategoryVariety(
          maxTimesPerWeek:
              _prefs.containsKey(_kVarietySnackMaxWeek)
                  ? toLimit(_kVarietySnackMaxWeek)
                  : 3,
          maxConsecutiveDays: toLimit(_kVarietySnackMaxConsec),
        ),
      },
    );
  }

  ProteinPowder? _buildProteinPowder() {
    final id = _prefs.getString(_kPowderIngredientId);
    final name = _prefs.getString(_kPowderName);
    if (id == null || name == null) return null;
    return ProteinPowder(
      ingredientId: id,
      name: name,
      proteinPer100g: _prefs.getDouble(_kPowderProteinPer100g) ?? 0.0,
      kcalPer100g: _prefs.getDouble(_kPowderKcalPer100g) ?? 0.0,
      gramsInStock: _prefs.getDouble(_kPowderGramsInStock) ?? 0.0,
      autoFillGap: _prefs.getBool(_kPowderAutoFill) ?? true,
      lowStockWarning: _prefs.getBool(_kPowderLowStockWarning) ?? true,
    );
  }

  ShoppingConfig _buildShoppingConfig() => ShoppingConfig(
        shoppingDays: _splitToSet(_prefs.getString(_kShoppingDays))
            .map((s) => int.tryParse(s))
            .whereType<int>()
            .toSet()
            .let((s) => s.isEmpty ? {7} : s),
        intervalWeeks: _prefs.getInt(_kShoppingInterval) ?? 1,
      );

  // ── Individual setters ────────────────────────────────────────────────────

  Future<void> setKcalTarget(double kcal) =>
      _prefs.setDouble(_kKcalTarget, kcal);

  Future<void> setProteinTarget(double? g) => g != null
      ? _prefs.setDouble(_kProteinTarget, g)
      : _prefs.remove(_kProteinTarget);

  Future<void> setFatTarget(double? g) => g != null
      ? _prefs.setDouble(_kFatTarget, g)
      : _prefs.remove(_kFatTarget);

  Future<void> setCarbsTarget(double? g) => g != null
      ? _prefs.setDouble(_kCarbsTarget, g)
      : _prefs.remove(_kCarbsTarget);

  Future<void> setAutoField(MacroField field) =>
      _prefs.setString(_kAutoField, field.firestoreName);

  Future<void> setMinKcal(double? v) =>
      v != null ? _prefs.setDouble(_kMinKcal, v) : _prefs.remove(_kMinKcal);

  Future<void> setMaxKcal(double? v) =>
      v != null ? _prefs.setDouble(_kMaxKcal, v) : _prefs.remove(_kMaxKcal);

  Future<void> setMinProtein(double? v) => v != null
      ? _prefs.setDouble(_kMinProtein, v)
      : _prefs.remove(_kMinProtein);

  Future<void> setMaxProtein(double? v) => v != null
      ? _prefs.setDouble(_kMaxProtein, v)
      : _prefs.remove(_kMaxProtein);

  Future<void> setMinFat(double? v) =>
      v != null ? _prefs.setDouble(_kMinFat, v) : _prefs.remove(_kMinFat);

  Future<void> setMaxFat(double? v) =>
      v != null ? _prefs.setDouble(_kMaxFat, v) : _prefs.remove(_kMaxFat);

  Future<void> setMinCarbs(double? v) => v != null
      ? _prefs.setDouble(_kMinCarbs, v)
      : _prefs.remove(_kMinCarbs);

  Future<void> setMaxCarbs(double? v) => v != null
      ? _prefs.setDouble(_kMaxCarbs, v)
      : _prefs.remove(_kMaxCarbs);

  Future<void> setDietTypes(Set<DietType> types) =>
      _prefs.setString(_kDietTypes, types.map((t) => t.firestoreName).join(','));

  Future<void> toggleDietType(DietType type) async {
    final current = _buildDietPreferences().dietTypes;
    final updated =
        current.contains(type) ? current.difference({type}) : {...current, type};
    await setDietTypes(updated);
  }

  Future<void> setAllergies(Set<AllergyType> allergies) =>
      _prefs.setString(
          _kAllergies, allergies.map((a) => a.firestoreName).join(','));

  Future<void> toggleAllergy(AllergyType allergy) async {
    final current = _buildDietPreferences().allergies;
    final updated = current.contains(allergy)
        ? current.difference({allergy})
        : {...current, allergy};
    await setAllergies(updated);
  }

  Future<void> setExcludedIngredients(Set<String> ids) =>
      _prefs.setString(_kExcludedIds, ids.join(','));

  Future<void> setPreferredIngredients(Set<String> ids) =>
      _prefs.setString(_kPreferredIds, ids.join(','));

  Future<void> setDislikedIngredients(Set<String> ids) =>
      _prefs.setString(_kDislikedIds, ids.join(','));

  Future<void> setVarietyLevel(VarietyLevel level) =>
      _prefs.setString(_kVarietyLevel, level.firestoreName);

  Future<void> setSnackOptionalFill(bool enabled) =>
      _prefs.setBool(_kSnackOptionalFill, enabled);

  Future<void> setShoppingDays(Set<int> days) =>
      _prefs.setString(_kShoppingDays, days.join(','));

  Future<void> setShoppingInterval(int weeks) =>
      _prefs.setInt(_kShoppingInterval, weeks);

  Future<void> setProteinPowder({
    required String ingredientId,
    required String name,
    required double proteinPer100g,
    required double kcalPer100g,
    required double gramsInStock,
  }) async {
    await _prefs.setString(_kPowderIngredientId, ingredientId);
    await _prefs.setString(_kPowderName, name);
    await _prefs.setDouble(_kPowderProteinPer100g, proteinPer100g);
    await _prefs.setDouble(_kPowderKcalPer100g, kcalPer100g);
    await _prefs.setDouble(_kPowderGramsInStock, gramsInStock);
  }

  Future<void> setPowderAutoFill(bool enabled) =>
      _prefs.setBool(_kPowderAutoFill, enabled);

  Future<void> setPowderLowStockWarning(bool enabled) =>
      _prefs.setBool(_kPowderLowStockWarning, enabled);

  Future<void> setTheme(AppTheme theme) =>
      _prefs.setString(_kTheme, theme.name);

  Stream<AppTheme> watchTheme() async* {
    // SharedPreferences doesn't have a native stream; yield current value.
    final s = _prefs.getString(_kTheme) ?? 'system';
    try {
      yield AppTheme.values.firstWhere((e) => e.name == s);
    } catch (_) {
      yield AppTheme.system;
    }
  }

  // ── MealSlotConfig ────────────────────────────────────────────────────────

  Future<void> setMealSlotConfig(int dayOfWeek, DaySlotConfig config) =>
      _dao.upsertMealSlotConfig(MealSlotConfigsTableCompanion(
        dayOfWeek: Value(dayOfWeek),
        hasBreakfast: Value(config.hasBreakfast),
        hasLunch: Value(config.hasLunch),
        hasDinner: Value(config.hasDinner),
        snackCount: Value(config.snackCount),
      ));

  // ── BatchCookingGroups ────────────────────────────────────────────────────

  Future<void> saveBatchCookingGroups(
      List<BatchCookingGroup> groups) async {
    await _dao.deleteAllBatchCookingGroups();
    for (int i = 0; i < groups.length; i++) {
      final g = groups[i];
      await _dao.upsertBatchCookingGroup(BatchCookingGroupsTableCompanion(
        id: Value('batch_$i'),
        category: Value(g.category.firestoreName),
        days: Value(g.days.join(',')),
        batchNumber: Value(g.batchNumber),
      ));
    }
  }

  // ── OptimizerRules ────────────────────────────────────────────────────────

  Future<void> upsertRule(OptimizerRule rule) {
    final id = rule.id.isEmpty ? _uuid.v4() : rule.id;
    return _dao.upsertOptimizerRule(OptimizerRulesTableCompanion(
      id: Value(id),
      type: Value(rule.type.firestoreName),
      target: Value(rule.target),
      targetName: Value(rule.targetName),
      constraint: Value(rule.constraint.firestoreName),
      value: Value(rule.value),
    ));
  }

  Future<void> deleteRule(String ruleId) =>
      _dao.deleteOptimizerRule(ruleId);

  Stream<List<OptimizerRule>> watchRules() =>
      _dao.watchOptimizerRules().map(
            (rows) => rows.map(_optimizerRuleFromRow).toList(),
          );

  // ── Helpers ───────────────────────────────────────────────────────────────

  Set<String> _splitToSet(String? s) {
    if (s == null || s.isEmpty) return {};
    return s.split(',').map((e) => e.trim()).where((e) => e.isNotEmpty).toSet();
  }
}

// ── Entity conversion ─────────────────────────────────────────────────────────

OptimizerRule _optimizerRuleFromRow(OptimizerRulesTableData row) =>
    OptimizerRule(
      id: row.id,
      type: RuleTargetType.fromFirestore(row.type),
      target: row.target,
      targetName: row.targetName,
      constraint: ConstraintType.fromFirestore(row.constraint),
      value: row.value,
    );

extension<T> on T {
  R let<R>(R Function(T) block) => block(this);
}
